### ISSUE TITLE
HostnameEndpoint leaves timer running when cancelled

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8223]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)| @warner reported|
|-|-|
|Trac ID|trac#8223|
|Type|defect|
|Created|2016-02-28 18:16:57Z|
|Branch|https://github.com/twisted/twisted/tree/hostname-endpoint-dirty-reactor-8223|

As noted in [#8014](https://github.com/twisted/twisted/issues/8014)#comment:15 , when a `HostnameEndpoint`'s Deferred is cancelled while waiting for a connection to occur, it leaves the `DelayedCall` running. While this might nott affect normal applications, any `trial` tests that hit this case will fail, with a "Dirty Reactor Error".

The attached patch fixes up the code to cancel the timer in all cases, and adds tests.

Attachments:

* [delayed-calls-8014.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5ab/5abf5b53c561c1d8ffb28c08fa318dbf1ed10982/8b7413bd9b51d9b525ede809183a3841374213fc.diff) (4254 bytes) - added by warner on 2016-02-28 18:17:11Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8223 8223
type__defect defect
reporter__warner warner
priority__normal normal
milestone__None None
branch__branches_hostname_endpoint_dirty_reactor_8223 branches/hostname-endpoint-dirty-reactor-8223
branch_author__mithrandi mithrandi
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1456683417722056 1456683417722056
changetime__1456779955045156 1456779955045156
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

Thanks for the patch.

You are listed as a commiter. Why not create a branch for your patch and send it to buildbot?

It needs a news file fragment.

---------

Minor comments

For the 

```
self.assertEqual([], mreactor.getDelayedCalls())
and 
self.assertEqual([], mreactor.getDelayedCalls())
```

maybe the test can be refactored with something like this


```
def getReactor():
    someReactor = setUpNewReactor()
    self.addCleanup(self.assertEqual,[], someReactor)
    return someReactor

```

in this way we make sure that each reactor used with be automatically checked.

The current method look error prone to me, as when writing a new test someone might forget to add the assertion.

For my project I have an `assertReactorIsClean` helper assertion which also checks for threads / readers and writers and is called on tearDown / cleanup

Maybe we can also have something like that in Twisted... if this does not already exists.

Leaving this for review as is Sunday evening at 23:30 here, not really a good time to do reviews.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_2"></a>|@mithrandi commented|
|-|-|

(In [46871]) Branching to hostname-endpoint-dirty-reactor-8223.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_3"></a>|@mithrandi commented|
|-|-|

Put this in a branch and spun the builders. I also added a .misc news fragment (I don't think this needs a mention in NEWS as the issue it fixes is unreleased as far as I can tell).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

I think this is a duplicate of [#8202](https://github.com/twisted/twisted/issues/8202)?

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)<a name="note_5"></a>|@warner commented|
|-|-|

glyph: I think this is different than [[#8202](https://github.com/twisted/twisted/issues/8202)](https://github.com/twisted/twisted/issues/8202). Imagine that you have two viable addresses for the given hostname. Then:

* #8202: if both connections fail, the `.connect()` Deferred errbacks 300ms after the last failure, but it should errback immediately after the last failure
* [#8223](https://github.com/twisted/twisted/issues/8223): if the `.connect()` Deferred is cancelled at any point before 300ms after the last failure, the `DelayedCall` is left running, but it should be stopped.

My patch should fix [#8223](https://github.com/twisted/twisted/issues/8223), but won't fix [#8202](https://github.com/twisted/twisted/issues/8202). The fix for that one will involve fetching the next endpoint from the iterator at the *end* of `iterateEndpoint()`, and terminating right away if it's None, instead of the current practice of fetching it at the *beginning* (think `do{}while()` instead of `while(){}`).

adirobian: oh, I didn't think I still had a commit bit. It's been so long since I've contributed anything to Twisted.. I don't know the current tooling or branch-creation process. I'll dig into the how-to-contribute docs and see if I can get SVN set up for the next patch. Thanks for creating the branch for me!

I agree that new tests are liable to not include that assertion. I noticed some tests for which asserting a clean reactor is *not* appropriate (where the test intentionally stops moving time forward part-way through the connection process), like `test_freeFunctionDeferToThread`, `test_defaultGAI`, and `test_nameResolution`. So I wasn't sure where a good common location would be.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_6"></a>|@mithrandi commented|
|-|-|

re: SVN, you may be interested in https://twistedmatrix.com/trac/wiki/GitMirror#ForTwistedcoredevelopers which I think most people much prefer to directly using SVN these days.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

I even recently updated GitMirror to have a quicker for-committers setup :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @glyph|
|-|-|

The tests look correct, I see they pass on the buildbot.  It needs a topfile, but otherwise, good to merge.  In deference to warner's forgetting how to drive our clunky toolchain I'll do the merge myself.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

oh, and it's even got a `.misc`.  Delightful.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set status to `closed`|
|-|-|

(In [46874]) Merge hostname-endpoint-dirty-reactor-8223: Clean up reactor after cancellation

Author: warner

Reviewer: glyph

Fixes: [#8223](https://github.com/twisted/twisted/issues/8223)

As noted in [#8014](https://github.com/twisted/twisted/issues/8014)#comment:15 , when a
HostnameEndpoint's Deferred is cancelled while waiting for a connection to
occur, it leaves the DelayedCall running. While this might not affect normal
applications, any trial tests that hit this case will fail, with a "Dirty
Reactor Error".
